### DISCUSSION TITLE
Query doesn’t support Jsonb + Rust Generic Type

### DISCUSSION THREAD

**Discussion by Moicen** [diesel-rs/diesel#3285] [Q&A]
I have a table structure like this:

```rust
#[derive(
AsChangeset,
Associations,
Identifiable,
PartialEq,
Serialize,
Deserialize,
Queryable,
Debug,
Insertable,
Clone,
)]
#[serde(bound = "")]
#[belongs_to(HtyTemplate, foreign_key = "template_id")]
#[belongs_to(HtyApp, foreign_key = "app_id")]
#[table_name = "hty_template_data"]
pub struct HtyTemplateData<T: Debug + Serialize + DeserializeOwned + Clone> {
    pub id: String,
    pub app_id: String,
    pub template_id: String,
    pub template_val: Option<String>,
    pub template_text: Option<SingleVal<T>>,
    pub created_at: NaiveDateTime,
    pub created_by: String,
}
```

It contains a generic type `template_text` which has generic type `Option<SingleVal<T>>`. The `SingleVal` is defined like this:

```rust
#[derive(AsExpression, FromSqlRow, Debug, Default, Serialize, Deserialize, PartialEq, Clone)]
#[sql_type = "Jsonb"]
#[serde(bound = "")]
pub struct SingleVal<T: Debug + Serialize + DeserializeOwned> {
    pub val: Option<T>,
}
```

It has PostgreSQL type `Jsonb`, and it’s `val` is defined as generic type `T`, and the `from_sql` and `to_sql` is defined like this:

```rust
impl<T: Debug + Serialize + DeserializeOwned> diesel::deserialize::FromSql<diesel::sql_types::Jsonb, diesel::pg::Pg> for SingleVal<T>
{
    fn from_sql(bytes: Option<&[u8]>) -> diesel::deserialize::Result<Self> {
        let value = <serde_json::Value as diesel::deserialize::FromSql<
            diesel::sql_types::Jsonb,
            diesel::pg::Pg,
        >>::from_sql(bytes)?;
        Ok(serde_json::from_value(value)?)
    }
}

impl<T: Debug + Serialize + DeserializeOwned> diesel::serialize::ToSql<diesel::sql_types::Jsonb, diesel::pg::Pg> for SingleVal<T> {
    fn to_sql<W: std::io::Write>(
        &self,
        out: &mut diesel::serialize::Output<W, diesel::pg::Pg>,
    ) -> diesel::serialize::Result {
        let value = serde_json::to_value(self)?;
        <serde_json::Value as diesel::serialize::ToSql<
            diesel::sql_types::Jsonb,
            diesel::pg::Pg,
        >>::to_sql(&value, out)
    }
}
```

And here is the usage of the generic type:

```rust
impl<T> HtyTemplateData<T> where
    T: Debug + Serialize + DeserializeOwned + Clone
{
    pub fn find_with_template_id_and_app_id(id_template: &String, id_app: &String, conn: &PgConnection) -> anyhow::Result<HtyTemplateData<T>> {
        debug!("find_with_template_id_and_app_id -> id_template: {:?}, id_app: {:?}", id_template, id_app);

        match hty_template_data::table
            .filter(hty_template_data::template_id.eq(id_template)
                .and(hty_template_data::app_id.eq(id_app)))
            .select(hty_template_data::all_columns)
            .first::<HtyTemplateData<String>>(conn)
        {
            Ok(data) => {
                let text = data.template_text.unwrap().val.unwrap();
                debug!("find_with_template_id_and_app_id -> template_text: {:?}", text);

                let json_text = serde_json::from_str::<T>(&text).unwrap();

                debug!("find_with_template_id_and_app_id -> json_text: {:?}", json_text);

                let resp_data = HtyTemplateData {
                    id: data.id.clone(),
                    app_id: data.app_id.clone(),
                    template_id: data.template_id.clone(),
                    template_val: data.template_val.clone(),
                    template_text: Some(SingleVal { val: Some(json_text.clone()) }),
                    created_at: data.created_at.clone(),
                    created_by: data.created_by.clone(),
                };

                debug!("find_with_template_id_and_app_id -> app: {:?}", resp_data);
                Ok(resp_data)
            }
            Err(e) => {
                debug!("find_with_template_id_and_app_id -> e: {:?}", e);
                Err({
                    anyhow!(HtyErr {code: HtyErrCode::DbErr,reason: Some(e.to_string())})
                })
            }
        }
    }
}
```

In the above implementation, I have to load the data by defining `Jsonb` field generic type as `String`, and then use `serde` to deserialize it to the target type manually:

![EFDCC3EC-B1B3-42A3-B241-3AD9AF553DFE](https://user-images.githubusercontent.com/3773928/185777735-f86141b1-ff3b-43d4-ae47-e114bb872b94.png)

However if I changed the select code to return the type directly like this:

![8A0483FF-CCEE-4D70-A8BD-6280B6944952](https://user-images.githubusercontent.com/3773928/185777744-dc1c1606-c691-49cb-8fe1-508c270615b5.png)

During the runtime it will throw error like this:

![327221661061030_ pic](https://user-images.githubusercontent.com/3773928/185777750-13f4f36f-93b1-4d4d-a76c-75d18d425133.jpg)

I’m not sure whether this is a usage problem or a diesel bug. Can anyone please provide help on this? Thanks!

**Comment by weiznich** [comment]
This is likely a problem in your implementation, since  I use similar implementations in my own code. Unfortunately it is not possible to tell whats exactly the problem there as your question is missing important details. Please provide at least information about your diesel version and your database schema. Ideally you are able to provide an minimal reproducible example that others could that to reproduce the described behavior. Such an minimal example makes it much easier to answer such questions.

**Reply by Moicen** [reply]
> This is likely a problem in your implementation, since I use similar implementations in my own code. Unfortunately it is not possible to tell whats exactly the problem there as your question is missing important details. Please provide at least information about your diesel version and your database schema. Ideally you are able to provide an minimal reproducible example that others could that to reproduce the described behavior. Such an minimal example makes it much easier to answer such questions.

Thanks for checking!

**Comment by Moicen** [comment]
@weiznich After upgrading the rust to latest nightly version and diesel to latest version, the problem disappears. Thank you for help!
